nav2_graceful_motion_controller/src/graceful_motion_controller.cpp

```diff
@@ -0,0 +1,335 @@
+  control_law_->setSpeedLimit(params_->v_linear_min, params_->v_linear_max, params_->v_angular_max);
+}
+
+geometry_msgs::msg::PoseStamped GracefulMotionController::getMotionTarget(
+  const double & motion_target_dist,
+  const nav_msgs::msg::Path & transformed_plan)
+{
+  // Find the first pose which is at a distance greater than the motion target distance
+  auto goal_pose_it = std::find_if(
+    transformed_plan.poses.begin(), transformed_plan.poses.end(), [&](const auto & ps) {
+      return std::hypot(ps.pose.position.x, ps.pose.position.y) >= motion_target_dist;
+    });
+
+  // If the pose is not far enough, take the last pose
+  if (goal_pose_it == transformed_plan.poses.end()) {
+    goal_pose_it = std::prev(transformed_plan.poses.end());
+  }
+
+  return *goal_pose_it;
+}
+
+bool GracefulMotionController::simulateTrajectory(
+  const geometry_msgs::msg::PoseStamped & robot_pose,
+  const geometry_msgs::msg::PoseStamped & motion_target, nav_msgs::msg::Path & trajectory)
+{
+  // Check for collision before moving
+  if (inCollision(
+      robot_pose.pose.position.x, robot_pose.pose.position.y,
+      tf2::getYaw(robot_pose.pose.orientation)))
+  {
+    return false;
+  }
+
+  // First pose
+  geometry_msgs::msg::PoseStamped next_pose;
+  next_pose.header.frame_id = costmap_ros_->getBaseFrameID();
+  next_pose.pose.orientation.w = 1.0;
+  trajectory.poses.push_back(next_pose);
+
+  double distance = std::numeric_limits<double>::max();
+  double resolution_ = costmap_ros_->getCostmap()->getResolution();
+  double dt = (params_->v_linear_max > 0.0) ? resolution_ / params_->v_linear_max : 0.0;
+
+  // Set max iter to avoid infinite loop
+  unsigned int max_iter = 2 * sqrt(
+    motion_target.pose.position.x * motion_target.pose.position.x +
+    motion_target.pose.position.y * motion_target.pose.position.y) / resolution_;
+
+  // Generate path
+  do{
+    // Apply velocities to calculate next pose
+    next_pose.pose = control_law_->calculateNextPose(dt, motion_target.pose, next_pose.pose);
+
+    // Add the pose to the trajectory for visualization
+    trajectory.poses.push_back(next_pose);
+
+    // Check for collision
+    geometry_msgs::msg::PoseStamped global_pose;
+    tf2::doTransform(next_pose, global_pose, costmap_transform_);
+    if (inCollision(
+        global_pose.pose.position.x, global_pose.pose.position.y,
+        tf2::getYaw(global_pose.pose.orientation)))
+    {
+      return false;
+    }
+
+    // Check if we reach the goal
+    distance = nav2_util::geometry_utils::euclidean_distance(motion_target.pose, next_pose.pose);
+  }while(distance > resolution_ && trajectory.poses.size() < max_iter);
+
+  return true;
+}
+
+geometry_msgs::msg::Twist GracefulMotionController::rotateToTarget(const double & angle_to_target)
+{
+  double sign = angle_to_target < 0.0 ? -1.0 : 1.0;
+  geometry_msgs::msg::Twist vel;
+  vel.linear.x = 0.0;
+  vel.angular.z = sign * params_->v_angular_max;
```

Comment:
I would recommend you actually use the angle_to_target rather than a fixed -1/+1 - and then have a scaling parameter - otherwise this almost sure to overshoot (you'll be going max_angular_v as you cross the goal, and you still need to decelerate)